fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/RecursiveUnionCursor.java

```diff
@@ -0,0 +1,264 @@
+                    }
+                } else {
+                    return wrapNextResult(recordCursorResult);
+                }
+            });
+        } else {
+            return onNextRecursive();
+        }
+    }
+
+    @Nonnull
+    private CompletableFuture<RecordCursorResult<T>> onNextRecursive() {
+        return recursiveCursor.onNext().thenCompose(recordCursorResult -> {
+            if (!recordCursorResult.hasNext()) {
+                if (recordCursorResult.getNoNextReason().isSourceExhausted()) {
+                    return recurse();
+                } else {
+                    return wrapLastResult(recordCursorResult);
+                }
+            } else {
+                return wrapNextResult(recordCursorResult);
+            }
+        });
+    }
+
+    @Nonnull
+    private CompletableFuture<RecordCursorResult<T>> recurse() {
+        if (recursiveStepCompletionCallback.get()) {
+            // restart the cursor of the recursive state, without plugging in any continuation from previous recursive
+            // state.
+            recursiveCursor = recursiveCursorSupplier.get();
+            return onNextRecursive();
+        }
+        return CompletableFuture.completedFuture(RecordCursorResult.exhausted());
+    }
+
+    @Nonnull
+    private CompletableFuture<RecordCursorResult<T>> wrapLastResult(@Nonnull RecordCursorResult<T> innerCursorResult) {
+        return CompletableFuture.completedFuture(RecordCursorResult.withoutNextValue(
+                new Continuation(isInitialState, isReadingFromInitialCursorSupplier,
+                        innerCursorResult::getContinuation, () -> RecordCursorStartContinuation.START, false),
+                innerCursorResult.getNoNextReason()));
+    }
+
+    @Nonnull
+    private CompletableFuture<RecordCursorResult<T>> wrapNextResult(@Nonnull RecordCursorResult<T> innerCursorResult) {
+        final var continuation = new Continuation(isInitialState, isReadingFromInitialCursorSupplier,
+                innerCursorResult::getContinuation, () -> RecordCursorStartContinuation.START, false);
+        return CompletableFuture.completedFuture(RecordCursorResult.withNextValue(innerCursorResult.get(), continuation));
+    }
+
+    @Override
+    public void close() {
+        initialCursor.close();
+        recursiveCursor.close();
+    }
+
+    @Override
+    public boolean isClosed() {
+        return recursiveCursor.isClosed(); // todo: improve.
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return executor;
+    }
+
+    @Override
+    public boolean accept(@Nonnull final RecordCursorVisitor visitor) {
+        if (visitor.visitEnter(this)) {
+            initialCursor.accept(visitor);
+            recursiveCursor.accept(visitor);
+        }
+        return visitor.visitLeave(this);
+    }
+
+    @Nonnull
+    public static <T> RecursiveUnionCursor<T> from(@Nullable final byte[] unparsed,
```

Comment:
This needs some Javadoc. In particular,

- `unparsed` isn't clear as to what it is; maybe `continuation` in its name would help as well.
- the sense of `recursiveStepCompletionCallback` isn't obvious: stop when `true` or continue when `true`

Reply:
I refactored the algorithm and added documentation.